# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama: Mohamad Arvin Fadriansyah
- IPK / Nilai Rapot: 3.10 / 77.5

In [1]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [2]:
# your code goes here!

# modal 20 pasang Sepatu
modal = 1000000

# kuantitas (1 kodi)
kuantitas = 20

# harga jual 1 sepatu
harga_jual = 80000

# Total pendapatan dari penjualan 20 sepatu
total_penjualan = kuantitas * harga_jual

profit = total_penjualan - modal
print("Total Profit: ",profit)




Total Profit:  600000


Jadi, dari profit di atas, dapat disimpulkan bahwa toko sepatu memiliki profit sebesar 600.000,00 rupiah

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut :
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya :
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [3]:
# your code goes here!
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
print(response)

<Response [200]>


In [4]:
response.content

b'{\n  "response":{"numFound":1571,"start":0,"docs":[\n      {\n        "id":"10.1371/journal.pone.0026564",\n        "abstract":["\\n        Little is known about the viruses infecting most species. Even in groups as well-studied as Drosophila, only a handful of viruses have been well-characterized. A viral metagenomic approach was used to explore viral diversity in 83 wild-caught Drosophila innubila, a mushroom feeding member of the quinaria group. A single fly that was injected with, and died from, Drosophila C Virus (DCV) was added to the sample as a control. Two-thirds of reads in the infected sample had DCV as the best BLAST hit, suggesting that the protocol developed is highly sensitive. In addition to the DCV hits, several sequences had Oryctes rhinoceros Nudivirus, a double-stranded DNA virus, as a best BLAST hit. The virus associated with these sequences was termed Drosophila innubila Nudivirus (DiNV). PCR screens of natural populations showed that DiNV was both common and wi

In [5]:
dataset = response.content

In [6]:
import json
response_in_dict=json.loads(dataset)
json.loads(dataset)

{'response': {'numFound': 1571,
  'start': 0,
  'docs': [{'id': '10.1371/journal.pone.0026564',
    'abstract': ['\n        Little is known about the viruses infecting most species. Even in groups as well-studied as Drosophila, only a handful of viruses have been well-characterized. A viral metagenomic approach was used to explore viral diversity in 83 wild-caught Drosophila innubila, a mushroom feeding member of the quinaria group. A single fly that was injected with, and died from, Drosophila C Virus (DCV) was added to the sample as a control. Two-thirds of reads in the infected sample had DCV as the best BLAST hit, suggesting that the protocol developed is highly sensitive. In addition to the DCV hits, several sequences had Oryctes rhinoceros Nudivirus, a double-stranded DNA virus, as a best BLAST hit. The virus associated with these sequences was termed Drosophila innubila Nudivirus (DiNV). PCR screens of natural populations showed that DiNV was both common and widespread taxonomic

In [7]:
for keys, value in response_in_dict.items():
    print("Keys dari dimensi pertama: ",keys)
    for key2, value2 in value.items():
        print("Keys dari dimensi kedua: ",key2)

Keys dari dimensi pertama:  response
Keys dari dimensi kedua:  numFound
Keys dari dimensi kedua:  start
Keys dari dimensi kedua:  docs


Jadi, dari url link di atas, kami menyimpulkan bahwa
Keys dari dimensi pertama:  response
Keys dari dimensi kedua:  numFound
Keys dari dimensi kedua:  start
Keys dari dimensi kedua:  docs

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c).

In [8]:
# your code goes here!
list_kota = ['Jakarta', 'Depok', 'Bogor', 'Bekasi', 'Tanggerang', 'Bandung', 'Padang', 'Medan', 'Palembang', 'Surabaya']
print("Total Kota: " ,len(list_kota))

#Nomor 1
list_kota_terurut = sorted(list_kota)
print("Kota Terurut dalam abjad: " ,list_kota_terurut)

#Nomor 2
print("4 kota pertama: " ,list_kota_terurut[:4])

#Nomor 3
print("3 kota terakhir: " ,list_kota_terurut[-3:])


Total Kota:  10
Kota Terurut dalam abjad:  ['Bandung', 'Bekasi', 'Bogor', 'Depok', 'Jakarta', 'Medan', 'Padang', 'Palembang', 'Surabaya', 'Tanggerang']
4 kota pertama:  ['Bandung', 'Bekasi', 'Bogor', 'Depok']
3 kota terakhir:  ['Palembang', 'Surabaya', 'Tanggerang']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa 
Kota Terurut dalam abjad:  ['Bandung', 'Bekasi', 'Bogor', 'Depok', 'Jakarta', 'Medan', 'Padang', 'Palembang', 'Surabaya', 'Tanggerang'] 
4 kota pertama:  ['Bandung', 'Bekasi', 'Bogor', 'Depok']
3 kota terakhir:  ['Palembang', 'Surabaya', 'Tanggerang']

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu:
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [9]:
# your code goes here!
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
#Nomor 1
print('Transkasi minimum: ',min(total_transaksi))

#Nomor 2
print('Rata-rata Transaksi: ', sum(total_transaksi)/len(total_transaksi))

#Nomor 3
print('Transkasi maksimal: ',max(total_transaksi))

Transkasi minimum:  10
Rata-rata Transaksi:  25.6
Transkasi maksimal:  45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal adalah 10, max adalah 45, rata-rata adalah 25.6

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut :
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [10]:
# your code goes here!
arr = np.array([1, 3, 5, 7, 9])
print("Array bilangan ganjil: ", arr)

#Nomor 1
arr = arr*10
print("Array setelah dikali 10: ", arr)

#Nomor 2
arr = arr/3
print("Array setelah dibagi 3: ", arr)

#Nomor 3
arr = arr * np.array([-45, 43, 3, 56, 5])
print("Array setelah dikali dengan [-45, 43, 3, 56, 5]: ", arr)

#Nomor 4
print("Index dengan nilai tertinggi:", arr.argmax())
print("Index dengan nilai terendah:", arr.argmin())



Array bilangan ganjil:  [1 3 5 7 9]
Array setelah dikali 10:  [10 30 50 70 90]
Array setelah dibagi 3:  [ 3.33333333 10.         16.66666667 23.33333333 30.        ]
Array setelah dikali dengan [-45, 43, 3, 56, 5]:  [-150.          430.           50.         1306.66666667  150.        ]
Index dengan nilai tertinggi: 3
Index dengan nilai terendah: 0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa,
Array bilangan ganjil:  [1 3 5 7 9]
Array setelah dikali 10:  [10 30 50 70 90]
Array setelah dibagi 3:  [ 3.33333333 10.         16.66666667 23.33333333 30.        ]
Array setelah dikali dengan [-45, 43, 3, 56, 5]:  [-150.          430.           50.         1306.66666667  150.        ]
Index dengan nilai tertinggi: 3
Index dengan nilai terendah: 0



# Terimakasih, salam dari Bogor